ambry-router/src/main/java/com/github/ambry/router/QuotaAwareOperationController.java

```diff
@@ -0,0 +1,193 @@
+        responseHandler, notificationSystem, routerMetrics, kms, cryptoService, cryptoJobHandler, time,
+        nonBlockingRouter);
+    this.nonBlockingRouter = nonBlockingRouter;
+    this.routerMetrics = routerMetrics;
+  }
+
+  @Override
+  protected void pollForRequests(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop) {
+    try {
+      List<RequestInfo> newRequestsToSend = new ArrayList<>();
+      pollNewRequests(newRequestsToSend, requestsToDrop);
+      addToRequestQueue(newRequestsToSend);
+      drainRequestQueue(requestsToSend);
+    } catch (Exception e) {
+      logger.error("Operation Manager poll received an unexpected error: ", e);
+      routerMetrics.operationManagerPollErrorCount.inc();
+    }
+  }
+
+  /**
+   * Poll for new requests from each of the operation managers, and initiate backgroud deletes.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the new requests to be sent out.
+   * @param requestsToDrop a list of correlation IDs that will contain the IDs for requests that the network layer
+   *                       should drop.
+   */
+  private void pollNewRequests(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop) {
+    putManager.poll(requestsToSend, requestsToDrop);
+    getManager.poll(requestsToSend, requestsToDrop);
+    nonBlockingRouter.initiateBackgroundDeletes(getBackGroundDeleteRequests());
+    clearBackGroundDeleteRequests();
+    deleteManager.poll(requestsToSend, requestsToDrop);
+    ttlUpdateManager.poll(requestsToSend, requestsToDrop);
+    undeleteManager.poll(requestsToSend, requestsToDrop);
+  }
+
+  /**
+   * Add the specified {@link List} of {@link RequestInfo}s to the reqeust queue.
+   * @param requestInfos {@link List} of {@link RequestInfo} objects.
+   */
+  private void addToRequestQueue(List<RequestInfo> requestInfos) {
+    for (RequestInfo requestInfo : requestInfos) {
+      QuotaResource quotaResource = requestInfo.getChargeable().getQuotaResource();
+      if (quotaResource == null) {
+        quotaResource = UNKNOWN_QUOTA_RESOURCE;
+      }
+      requestQueue.putIfAbsent(quotaResource, new LinkedList<>());
+      requestQueue.get(quotaResource).add(requestInfo);
+    }
+  }
+
+  /**
+   * Drain the request queue based on quota and system resources and update the requests to be send to {@code requestsToSend}.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the requests to be sent out.
+   */
+  private void drainRequestQueue(List<RequestInfo> requestsToSend) {
+    drainQuotaCompliantRequests(requestsToSend);
+
+    List<QuotaResource> quotaResources = new ArrayList<>(requestQueue.keySet());
+    Collections.shuffle(quotaResources);
+    boolean done = false;
+    while (!requestQueue.isEmpty()) {
+      for (QuotaResource quotaResource : quotaResources) {
+        RequestInfo requestInfo = requestQueue.get(quotaResource).getFirst();
```

Comment:
Like we discussed offline, just noting that charged and allowed replica requests of the same chunk could be split  across `send` cycles since we are choosing only one replica request in each iteration. Thanks for planning to optimize it in subsequent PRs.

Reply:
shared the ticket you separately to track this work.